File: lib/models/application.py

class Application:
    def __init__(self, raw):
        self.name = raw['name']
        self.port = raw['port']

        self.image = {
            'uri': raw['image'],
            'version': raw['version']
        }

        if 'imagePullSecret' in raw:
            self.image['imagePullSecret'] = raw['imagePullSecret']

        self.ingress = None
        if 'url' in raw:
            parts = raw['url'].split('://', 1)

            self.ingress = {
                'tls': True if parts[0] == 'https' else False,
                'url': parts[1]
            }

        self.service = None
        if 'port' in raw:
            self.service = {
                'port': 80,
                'targetPort': raw['port']
            }

        self.env = list()
        if 'environment' in raw:
            for key, value in raw['environment'].items():
                self.env.append({
                    'name': key,
                    'value': value
                })

        self.replicas = raw.get('replicas', 1)

File: pkg/api/application.go
package api

import (
	"bytes"
	"fmt"
	"io"
	"sigs.k8s.io/yaml"
)

type IngressConfig struct {
	Annotations map[string]string
}

type Application struct {
	Name string
	Namespace string

	Image string
	Version string
	ImagePullSecret string

	Url string
	Port int32

	Replicas int32

	Environment map[string]string
	Volumes []map[string]string

	Ingress IngressConfig
}

func ParseApplication(reader io.Reader) (Application, error) {
	var (
		app Application
		buf bytes.Buffer
	)

	_, err := io.Copy(&buf, reader)
	if err != nil {
		return Application{}, fmt.Errorf("error reading stdin: %w", err)
	}

	err = yaml.Unmarshal(buf.Bytes(), &app)
	if err != nil {
		return Application{}, fmt.Errorf("error parsing yaml: %w", err)
	}

	return app, nil
}
